dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EventPersister.java

```diff
@@ -172,63 +169,94 @@ private void handleDataValues(
     payloadDataValues.forEach(
         dv -> {
           DataElement dataElement = preheat.getDataElement(dv.getDataElement());
-          checkNotNull(
-              dataElement,
-              "Data element should never be NULL here if validation is enforced before commit.");
 
           // EventDataValue.dataElement contains a UID
           EventDataValue eventDataValue = dataValueDBMap.get(dataElement.getUid());
 
-          ValuesHolder valuesHolder = getAuditAndDateParameters(eventDataValue, dv);
+          if (isNewDataValue(eventDataValue, dv)) {
```

Comment:
LGTM, the only thing is that this mapper method got very long, I'd split it in different methods, maybe each change log type condition (CREATE, UPDATE and DELETE) could be done different method

Reply:
Done